### PR TITLE
Add DragonFlyBSD support for phobos

### PR DESCRIPTION
**Pull request by dkgroot**
Related: 
- [DMD Pull Request](https://github.com/dlang/dmd/pull/7463)
- [DRuntime Pull Request](https://github.com/dlang/druntime/pull/1999)

Bootstrapped via 'dmd-cxx' branch: [See](https://github.com/dkgroot/dragonflybsd_dmd_port)

Notes:
- Exclude assertion for dragonfly in std/experimental/allocator/mallocator.d (Needs further investigation, not sure why aligned_realloc is causing trouble).

### CODE REVIEW

File: posix.mak [hunk 1]
```diff
@@ -25,9 +25,10 @@
 ################################################################################
 # Configurable stuff, usually from the command line
 #
-# OS can be linux, win32, win32wine, osx, or freebsd. The system will be
+# OS can be linux, win32, win32wine, osx, freebsd, netbsd or dragonflybsd. The system will be
 # determined by using uname
 
+#QUIET:=@
```
Comment (wilzbach) [inline posix.mak]: Any reason for this to be added?

File: posix.mak [hunk 2]
```diff
@@ -25,9 +25,10 @@
 ################################################################################
 # Configurable stuff, usually from the command line
 #
-# OS can be linux, win32, win32wine, osx, or freebsd. The system will be
+# OS can be linux, win32, win32wine, osx, freebsd, netbsd or dragonflybsd. The system will be
```
Comment (wilzbach) [inline posix.mak]: Nit: break it at 80 chars here (i.e. after dragonflybsd).

File: std/conv.d [hunk 3]
```diff
@@ -5519,7 +5519,7 @@ pure nothrow @safe @nogc unittest
 }
 
 // Test attribute propagation for UDTs
-pure nothrow @safe /* @nogc */ unittest
+pure nothrow @safe /+@nogc+/ unittest
```
Comment (wilzbach) [inline std/conv.d]: Unrelated change - remove?

File: std/math.d [hunk 4]
```diff
@@ -7444,6 +7444,34 @@ private real polyImpl(real x, in real[] A) @trusted pure nothrow @nogc
                 ;
             }
         }
+        else version (DragonFlyBSD)
+        {
+            asm pure nothrow @nogc // assembler by W. Bright
```
Comment (wilzbach) [inline std/math.d]: For the record: I'm not a huge fan of copying blocks as this means a bug fix will very likely miss one of these bits.
Anyhow as far as I know is Walter a huge proponent of using `version(A)` and not of `static if( A || B)`

File: std/parallelism.d [hunk 5]
```diff
@@ -984,6 +988,10 @@ uint totalCPUsImpl() @nogc nothrow @trusted
         {
             auto nameStr = "hw.ncpu\0".ptr;
         }
+        else version(DragonFlyBSD)
+        {
+            auto nameStr = "hw.ncpu\0".ptr;
```
Comment (wilzbach) [inline std/parallelism.d]: Off topic: this looks like a very old style. IIRC are strings literals always zero-terminated?

File: std/math.d (hunk 4 above)
Comment (dkgroot) [reply std/math.d]: I completely agree. I did see this 'argument' pop up several time in different PR's and Forum discussions. I even wanted to suggest a generic 'BSD' version flag, as they all have common origins, similar to the 'Posix' version. But i did not want to reopen this discussion.

File: std/conv.d (hunk 3 above)
Comment (dkgroot) [reply std/conv.d]: Made it easier to exclude ranges of unittests, while trying to track down a particular issue.
But reverting the change.

File: std/parallelism.d (hunk 5 above)
Comment (dkgroot) [reply std/parallelism.d]: Don't know, maybe someone else can answer this one.

File: posix.mak (hunk 1 above)
Comment (dkgroot) [reply posix.mak]: It was needed while debugging, accidentally committed. Retracted now.

File: std/parallelism.d (hunk 5 above)
Comment (nemanja-boric-sociomantic) [reply std/parallelism.d]: Yes, that's true, the string literals are always zero-terminated. I was wondering if something is also doing `nameStr.length` in which case it makes sense putting this `\0` but it doesn't look like it.

Comment (dkgroot) [reply std/parallelism.d]: @nemanja-boric-sociomantic 
So should i rewrite this to: ```auto nameStr = "hw.ncpu".ptr;```, or just leave it as is.

The surrounding lines are doing the same, though. and i do not want to correct them (Only DragonFlyBSD related lines should be touched in this PR).

Comment (nemanja-boric-sociomantic) [reply std/parallelism.d]: I would leave them as is IMHO and do this separately.

File: std/math.d (hunk 4 above)
Comment (ibuclaw) [reply std/math.d]: Or maybe we do away with the asm implementation and implement a better poly algorithm.

https://issues.dlang.org/show_bug.cgi?id=12084

Comment (BBasile) [reply std/math.d]: It's possible too to put the implementation in a token string  `enum polyAsm = q{ asm pure nothrow @nogc { ... } };` and to mix it

Comment (dkgroot) [reply std/math.d]: I think this discussion should be had outside this PR. Sounds more like a structural (ie ASM) discussion, that pertaining to this particular PR (correct).

File: std/experimental/allocator/mallocator.d [hunk 6]
```diff
@@ -375,6 +375,7 @@ version(Posix)
     AlignedMallocator.instance.alignedReallocate(c, 32, 32);
     assert(c.ptr);
 
+    version (DragonFlyBSD) {} else                    /* FIXME: assertion below fails, have not been able to figure out why, yet */
     assert(!AlignedMallocator.instance.alignedReallocate(c, size_t.max, 4096));
```
Comment (ZombineDev) [inline std/experimental/allocator/mallocator.d]: If this assert fails it follows that `alignedReallocate` successfully called `posix_memalign`, allocating more virtual memory than what could be reasonably available on the system. Can you try to debug this test and see what exactly breaks down?

Comment (dkgroot) [reply std/experimental/allocator/mallocator.d]: Reported this error to dragonfly ([dragonfly bug report](https://bugs.dragonflybsd.org/issues/3114))
See [PR Discussion](https://github.com/dlang/druntime/pull/1999#discussion_r157536030)
Will update this entry to add both these links, so the issue can be tracked.

Comment (ZombineDev) [reply std/experimental/allocator/mallocator.d]: Thanks!

File: std/regex/internal/parser.d [hunk 7]
```diff
@@ -1033,6 +1033,7 @@ if (isForwardRange!R && is(ElementType!R : dchar))
 {//@@@BUG@@@ write is @system
     with(zis)
     {
+        import std.regex.internal.kickstart;
```
[the remark this answers is not in the document]
Comment (dkgroot) [reply std/regex/internal/parser.d]: I did not know i added this.  I think this might be wrongfully merged in. Thanks for noticing.
I will remove it and re-run my tests.

[the remark this answers is not in the document]
Comment (dkgroot) [reply std/regex/internal/parser.d]: I had some trouble in line 1095 when running without this added import. I will recheck.

**Comment by dkgroot** [comment]
@ZombineDev Thanks for Reviewing BTW
@jmdavis Thanks !

File: std/regex/internal/parser.d (hunk 7 above)
[the remark this answers is not in the document]
Comment (dkgroot) [reply std/regex/internal/parser.d]: Looks like the issue was resolved (possible/maybe by the git rebase i did before committing the PR).
Will remove the import

**Comment by dkgroot** [comment]
@ZombineDev Thanks for the review. I guess i will still have to wait a little, to allow other to review, before squashing, right ? How do you guys normally like to see this squash done, using 'rebase -i HEAD~5'?

**Comment by ibuclaw** [comment]
@dkgroot. Yes, git rebase and then fix up all commits into one.  Amending the original commit message if necessary.

**Comment by ZombineDev** [comment]
I think the PR is good to go, so we don't need to wait more. I prefer doing squashing with `git rebase -i HEAD~<number of commits on top of master>` (in your case `git rebase -i HEAD~8`).

**Comment by dkgroot** [comment]
@ZombineDev I did the `git rebase -i HEAD~8` as you suggested, but now this PR includes other/unrelated commits (ie: #a11f63ff,#4a1065e1,#1ad40c60,#cae4a305), [See](https://github.com/dlang/phobos/pull/5941/commits). I hope this is not going to cause issues.

**Comment by nemanja-boric-sociomantic** [comment]
Just do `git fetch upstream` followed by `git rebase -i upstream/master`, given that the `upstream` is the name of the blessed remote.

**Comment by ZombineDev** [comment]
@dkgroot  Yes, as @nemanja-boric-sociomantic said, first rebase your branch on top of master and then squash it.

**Comment by nemanja-boric-sociomantic** [comment]
I see you lost all of your commits :-). `git reflog` will save you in case you don't know it already.

**Comment by andralex** [comment]
@dkgroot what I do when I have unrelated commits in my PR is I remove them during `git rebase -i`

**Comment by dkgroot** [comment]
Damn i completely messed it up. Will need a little time to figure this out.

**Comment by andralex** [comment]
@dkgroot you may want to send me email (see http://erdani.com/index.php/contact) and I'll invite you to https://dlang.slack.com. Then you can get community help in real time via text chat in the browser.

**Comment by dkgroot** [comment]
Did a 'git reset --hard ca27cdfec' which was my last known good state.
Will have to go a little slower this time.

**Comment by dkgroot** [comment]
I hope I managed to do better this time.

**Comment by ibuclaw** [comment]
@andralex - He is already on slack. :-)

**Comment by dkgroot** [comment]
Sorry for the 'git mess' I created. Accoding to @geod24 it is merge able as is, so leaving it alone for now.
Thanks for all your help guys

**Comment by andralex** [comment]
Thanks for the work and many thanks to the reviewers! I'll do the honors unless someone beats me to it.

**Comment by andralex** [comment]
Oh, was too hasty - many commits. Will undo this.

**Comment by dkgroot** [comment]
Thanks a lot everyone for the help / review / git stuff and the final merge !

**Comment by MartinNowak** [comment]
Thanks for your contribution.

> Oh, was too hasty - many commits. Will undo this.

Well a missing squash isn't the end of the world, it's only a single merge commit on master after all.
